### DISCUSSION TITLE
MappedAsDataclass many-to-many keep constructor flexibility

### DISCUSSION THREAD

**Discussion by EugeneLagonski** [sqlalchemy/sqlalchemy#9280] [Usage Questions] [answered]
Hello

I have ORM dataclass models connected with m2m relationship. I am trying to figure out how to create instances of a m2m secondary table model using ids of related and getting `AssertionError: Dependency rule tried to blank-out primary key column 'article_tag.tag_id' on instance '<ArticleTag at 0x1033bd600>'` exception.

Code snippet:
```python
class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Article(Base):
    __tablename__ = "article"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str]

    tags: Mapped[list["ArticleTag"]] = relationship(default_factory=list, back_populates="article")


class Tag(Base):
    __tablename__ = "tag"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str]

    relations: Mapped[list["ArticleTag"]] = relationship(default_factory=list, back_populates="tag")


class ArticleTag(Base):
    __tablename__ = "article_tag"
    article_id: Mapped[int] = mapped_column(ForeignKey("article.id"), primary_key=True, default=None)
    tag_id: Mapped[int] = mapped_column(ForeignKey("tag.id"), primary_key=True, default=None)
    article: Mapped[Article] = relationship(default=None)
    tag: Mapped[Tag] = relationship(default=None)


engine = create_engine("sqlite:///test.db", echo=True)
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

session = Session(engine, expire_on_commit=True)

tag1 = Tag(name="Tag1")
tag2 = Tag(name="Tag2")
session.add_all([tag1, tag2])
session.commit()

article = Article(
    name="Article",
    tags=[
        ArticleTag(tag_id=tag1.id),
        ArticleTag(tag_id=tag2.id),
    ],
)
session.add(article)
session.commit()
```

I know 3 ways to overcome this:
1. Pass the `Tag` instance into the `ArticleTag` constructor.
2. Replace the `ArticleTag.tag` relationship `default=None` setting to `init=False`
3. Configure all foreign keys and relationship fields as `init=False` and set them explicitly after object creation
4. In the process of writing this text I found one more way to fix this. I've added the deletion of relationship fields if they are `None` by `__post_init__` method to the `ArticleTag` class.
```python
    def __post_init__(self):
        if self.article is None:
            delattr(self, "article")
        if self.tag is None:
            delattr(self, "tag")
```

Is it possible to keep the flexibility of passing what is more convenient to form a relationship in this case? 

SQLAlchemy version: 2.0.3
Python version: 3.10.6

**Comment by zzzeek** [comment, marked as answer]
a novel problem to be sure.  Dataclasses intend to have this very specific kind of constructor that does not allow for arguments to be omitted.     One idea is to propose some new kind of constant that SQLAlchemy would recognize as "dont do anything", but this would not be simple to add and would also start to dilute what the semantics of the dataclass constructor actually means.

it seems more like what you really want is to just have the usual SQLAlchemy keyword constructor, so you can have that while keeping everything else dataclass by just turning off the dataclass constructor.  this works pretty nicely


```py
class ArticleTag(Base, init=False):
    __tablename__ = "article_tag"
    article_id: Mapped[int] = mapped_column(
        ForeignKey("article.id"), primary_key=True, default=None
    )
    tag_id: Mapped[int] = mapped_column(
        ForeignKey("tag.id"), primary_key=True, default=None
    )
    article: Mapped[Article] = relationship(default=None)
    tag: Mapped[Tag] = relationship(default=None)
````

**Reply by EugeneLagonski** [reply]
Thanks! It really fixes the problem.

**Reply by ar-work-acc** [reply]
Hi zzzekk, Eugene, 

I recently run into the same problem. 

From your reply, it seems that you cannot keep the `__init__` method generated from dataclass integration and at the same time, also keep both ways of initialization in the constructor without some manual hack? That is, using both `ArticleTag(tag=<Tag object>)` and `ArticleTag(tag_id=<some ID>)` to intialize an `ArticalTag` object.

I can do it my own way, but I just want to be sure.
